File: 31. Next Permutation.c
int cmpfunc(const void* a, const void* b){
    int x,y;
    
    memcpy(&x, a, sizeof(int));
    memcpy(&y, b, sizeof(int));
    
    return (x-y);    
}

void swap(int* nums, int i, int j){
    nums[i]^=nums[j];
    nums[j]^=nums[i];
    nums[i]^=nums[j];
}

void dumpArr(int* nums, int numsSize){
    int i;
    
    printf("ARRAY: ");
    for(i=0; i<numsSize; i++){
        printf("%d ", nums[i]);
    }
    printf("\n");
}

void nextPermutation(int* nums, int numsSize) {
    int max=-1, max_idx;
    int min=-1, min_idx;
    int i;
    
    for(i=numsSize-1; i>=0; i--){
        if(nums[i] >= max){
            max = nums[i];
            max_idx = i;
        }else{
            break;
        }
    }
    //printf("max_idx=%d, max=%d\n", max_idx, max);
    
    
    if(max_idx==0){
        //sort nums
        qsort(nums, numsSize, sizeof(int), cmpfunc);
    }else{
        for(i=numsSize-1; i>=max_idx; i--){
            if(nums[i] > nums[max_idx-1] && (min==-1 || min > nums[i])){
                min = nums[i];
                min_idx = i;
            }
        }        
        //printf("min_idx=%d, min=%d\n", min_idx, min);
        dumpArr(nums, numsSize);
        //swap nums[max_idx-1] & nums[min]
        swap(nums, max_idx-1, min_idx);
        
        //dumpArr(nums, numsSize);
        
        //sort
        qsort(nums+max_idx, numsSize-max_idx, sizeof(int), cmpfunc);
        
        //dumpArr(nums, numsSize);
    }
    
}

File: 31. Next Permutation.py
class Solution:
    def swap(self, l, i, j):
        if i != j:
            l[i]^=l[j]
            l[j]^=l[i]
            l[i]^=l[j]
        
    def reverse(self, l):
        for i in range(len(l)//2):
            self.swap(l, i, len(l)-i-1)
    
    def qSortPart(self, a, l, r, pivot):
        self.swap(a, pivot, r)
        
        centerIdx = l
        for i in range(l, r):
            if a[i] < a[r]:
                self.swap(a,i,centerIdx)
                centerIdx +=1
        
        self.swap(a, centerIdx, r)
        return centerIdx
        
    def qSort(self, a, l, r):
        if l<r:
            privot = self.qSortPart(a, l, r, l+(r-l)//2)
            self.qSort(a, l, privot-1)
            self.qSort(a, privot+1, r)
    

        
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        maxIdx = None
        
        for i in range(len(nums)-1, 0-1, -1):
            if maxIdx == None or nums[maxIdx] <= nums[i]:
                maxIdx = i
            else:
                break
                
        if maxIdx != None:
            if maxIdx == 0:
                #3,2,1 -> 1,2,3
                self.reverse(nums)
            else:
                rightMinIdx = None
                for i in range(len(nums)-1, maxIdx-1, -1):
                    if nums[i] > nums[maxIdx-1] and (rightMinIdx==None or nums[rightMinIdx] > nums[i]):
                        rightMinIdx = i
                
                if rightMinIdx != None:
                    self.swap(nums, maxIdx-1, rightMinIdx)
                    self.qSort(nums, maxIdx, len(nums)-1)
                        
